### PR TITLE
feat: replace dispatcher

### LINKED ISSUE #10920: Implement a many-to-one `Sequencer` that assigns record positions

### LINKED ISSUE #10921: `LogStorageAppender` should use `Sequencer` and `Serializer` instead of `Dispatcher`

### PR DESCRIPTION
**Pull request by lenaschoenburg**
This replaces the dispatcher with a `Sequencer`, a multi-producer single-consumer queue that assigns positions to all queue entries.

closes #10920 
closes #10921

### CODE REVIEW

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java
Comment (npepinpe) [inline logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java:147]: :x: Same as above - or am I missing something? :thinking: Is it because if the queue is full we want to ensure the consumer is triggered to fetch from it?

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java
Comment (npepinpe) [inline logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java:191]: :x: Shouldn't we signal only if the queue is not empty? I guess it doesn't hurt anyway :shrug: The other option is to have the consumer register itself and immediately poll for work :shrug:

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java
Comment (npepinpe) [inline logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java:194]: :thought_balloon: I think in the future we'll want to pass an interface, if only for testability, but let's see :+1:

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/serializer/DataFrameDescriptor.java
Comment (npepinpe) [inline logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/serializer/DataFrameDescriptor.java:16]: :upside_down_face: Maybe make it more explicit that this is kept around purely for historical reasons, and we hope to remove it :shrug:

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java
Comment (npepinpe) [inline logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java:63]: :thought_balloon: This logic is kind of tied to the serializer, in a way. I wonder if we could somehow centralize this? It's not a big deal though if we can't.

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/serializer/SequencedBatchSerializer.java
Comment (npepinpe) [inline logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/serializer/SequencedBatchSerializer.java:40]: :wrench: Don't thin this comment is necessary, the header already explains why we're framing/aligning as the old dispatcher. If we want to make it more explicit, then I would propose explaining the logic there (e.g. "The batch is serialized first by writing out each entry, framed and aligned, and then aligning the final result. This is required for backwards compatibility with the old Dispatcher module.").

File: logstreams/src/test/java/io/camunda/zeebe/logstreams/log/LogStreamBatchWriterTest.java
Comment (npepinpe) [inline logstreams/src/test/java/io/camunda/zeebe/logstreams/log/LogStreamBatchWriterTest.java:240]: :thought_balloon: I don't know if we should remove it from the API tests. In the end, it's part of the contract that if I write something, when I read it back, the timestamp will be there, no? I understand there's some overlap with the serialization process, but that's implementation details, I think.

File: logstreams/src/test/java/io/camunda/zeebe/logstreams/log/LogStreamWriterTest.java
Comment (npepinpe) [inline logstreams/src/test/java/io/camunda/zeebe/logstreams/log/LogStreamWriterTest.java:215]: :thought_balloon: Same as above

File: logstreams/src/test/java/io/camunda/zeebe/logstreams/log/LogStreamWriterTest.java
Comment (npepinpe) [inline logstreams/src/test/java/io/camunda/zeebe/logstreams/log/LogStreamWriterTest.java:283]: :question: Is this not possible anymore? I guess this would still be part of the contract? Though I'm all for removing it from the contract in general

**Review by npepinpe — COMMENTED** [review]
:+1: 

Sorry, it's a lot of comments, but most are just thoughts/suggestions. I also didn't see any new unit/integration tests, so I guess I'll hold off before commenting on missing test cases.

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LoggedEventImpl.java
Comment (oleschoenburg) [reply logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LoggedEventImpl.java:119]: Yeah, I'm annoyed by this too. It's just that my IntelliJ like to auto-import those and I didn't feel like fighting with it so far.

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java
Comment (oleschoenburg) [reply logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java:97]: We definitely want to do notify the consumer because the only case where nothing is enqueued is when the queue is full.

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java
Comment (oleschoenburg) [reply logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java:147]: Yes, that's exactly it.

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java
Comment (oleschoenburg) [reply logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java:191]: My reasoning was "doesn't hurt" but of course we could flip it and let the appender start polling. Does it matter?

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogStorageAppender.java
Comment (npepinpe) [reply logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogStorageAppender.java:170]: Where does it still know that every position is incremented by 1? :thinking: I guess I missed that. I thought now it just cared about highest/lowest, and that was it.

Comment (oleschoenburg) [reply logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogStorageAppender.java:170]: https://github.com/camunda/zeebe/blob/e08a42fdb7c6720a73cdde114f4c9233a22fab1a/logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/serializer/SequencedBatchSerializer.java#L37

Comment (oleschoenburg) [reply logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogStorageAppender.java:170]: So yeah, not technically the appender but the serializer.

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogStreamImpl.java
Comment (oleschoenburg) [reply logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogStreamImpl.java:174]: Yes, see here: https://github.com/camunda/zeebe/issues/11129

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogStorageAppender.java
Comment (oleschoenburg) [reply logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogStorageAppender.java:118]: Looks like it just adds a job and a subscription to the actor task. As long as the subscription exists, the task is resubmitted (see `allPhaseSubscriptionsTriggered`). Sounds to me like this is okay to do.

File: broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/BackupApiRequestHandlerStepTest.java
Comment (npepinpe) [reply broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/steps/BackupApiRequestHandlerStepTest.java:42]: All good, I think you answered this below

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogStorageAppender.java
Comment (npepinpe) [reply logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogStorageAppender.java:170]: I think that's alright, I group the Sequencer/SequencedBatch together imo as a single "unit"

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java
Comment (npepinpe) [reply logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java:191]: I think I would prefer the consumer to poll immediately, if only to avoid extra work of going through the actor scheduler via signaling. Once registered, the consumer WILL be signaled if new work comes in, so if it polls immediately and nothing's there, it won't lose anything (I think? :sweat_smile:)

But super minor, so let's make this a soft suggestion :smile: This approach is definitely more defensive, so I'm fine with either :+1:

File: logstreams/src/test/java/io/camunda/zeebe/logstreams/log/LogStreamBatchWriterTest.java
Comment (npepinpe) [reply logstreams/src/test/java/io/camunda/zeebe/logstreams/log/LogStreamBatchWriterTest.java:240]: I saw those tests more as testing the write API of the logstream module as a whole, not the specific writer. But I guess they were sort of specifically testing this interface :smile: 

All good :+1:

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LoggedEventImpl.java
Comment (oleschoenburg) [reply logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LoggedEventImpl.java]: Adjusted to not import static methods now

File: logstreams/src/test/java/io/camunda/zeebe/logstreams/log/LogStreamWriterTest.java
Comment (npepinpe) [reply logstreams/src/test/java/io/camunda/zeebe/logstreams/log/LogStreamWriterTest.java:283]: Considering LogAppendEntry is an interface (so can't guarantee it will always be non-null), I guess we need to do it in `tryWrite`.

:thought_balloon: I guess if we start having more validation logic in try write, then we can revisit the concept of splitting writer out of sequencer :shrug:

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java
Comment (oleschoenburg) [reply logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java]: :+1: I implemented your suggestion.

File: logstreams/src/test/java/io/camunda/zeebe/logstreams/log/LogStreamWriterTest.java
Comment (oleschoenburg) [reply logstreams/src/test/java/io/camunda/zeebe/logstreams/log/LogStreamWriterTest.java]: Added some basic validation: [a592140](https://github.com/camunda/zeebe/pull/11165/commits/a592140c0e2116de6f6d670b1a786069cb11badf)

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java
Comment (oleschoenburg) [reply logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java:90]: > So I would propose enabling fairness
Even though `ReentrantLock` points out lower throughput and limited effectiveness because thread scheduling is not fair? I think it might be worth a try but I'm not sure we could even tell if this has any effect or even if we have starved writer threads at all.

> and also locking interruptibly
I'm not sure I understand what we would get from that. Is anything in Zeebe ever interrupting writer threads? Or is it just a matter of "in theory this is better"?

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogStorageAppender.java
Comment (npepinpe) [inline logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogStorageAppender.java:120]: :upside_down_face: Could also just call `tryWriteBatch` directly on `onActorStarted`. The submitted call back is not called anyway until the actor is started, and we have the started hook for these kind of things.

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LoggedEventImpl.java
Comment (npepinpe) [inline logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LoggedEventImpl.java:139]: :question: Why remove the `toString` method? It's nice to have a string representation. Was it causing any issues?

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java
Comment (npepinpe) [reply logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java:90]: It's mostly a defensive technique, when doing asynchronous programming you want to avoid possible blocking points. In the case of a lock, it's mostly about deadlocks (though high latency variance between task execution isn't great either), which hopefully you never run into, but this is about preventing this issue (e.g. someone changes some other usage and we end up with this being blocking).

Anyway, as we discussed, I'd rather invest time in removing the lock entirely :+1:

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java
Comment (npepinpe) [inline logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java:116]: :upside_down_face: I don't notice changes between this and the interface docs, so we can just use the `{@inheritDoc}` tag, I think.

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java
Comment (npepinpe) [inline logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java:177]:
```suggestion

  /**
   * Retrieves, but does not remove, the first item in the sequenced batch queue.
   *
   * @return A {@link SequencedBatch} or null if none is available
   */
```

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java
Comment (npepinpe) [inline logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java:35]: :wrench: I think in general this class and most of its methods could have lowered visibility to package private. Lower/weaker visibility is useful to hint how your class should be used by consumers.

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/serializer/DataFrameDescriptor.java
Comment (npepinpe) [inline logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/serializer/DataFrameDescriptor.java:16]:
```suggestion
 * rest of the 12 bytes are unused. This is kept for backwards-compatibility reasons because {@link
```

File: logstreams/src/test/java/io/camunda/zeebe/logstreams/log/SequencerTest.java
Comment (npepinpe) [inline logstreams/src/test/java/io/camunda/zeebe/logstreams/log/SequencerTest.java:25]: :x: Please add tests to verify the sequencing of the batches is correct.

**Review by npepinpe — CHANGES_REQUESTED** [review]
:rocket: 

Very cool to see! Looking forward to the -6k lines PR ;)

- :x: Please add a unit test for the sequenced batch's serialization, specifically verifying that the framing/aligning is what we expect (such that it will be backwards compatible). As it's also the one setting the timestamp, you could add a test for that here instead of the writer as well, as you prefer. I would tentatively also add a test about the byte ordering, but I guess this is indirectly covered by the writer/reader tests, so can be omitted - it narrows the scope a little, but not by much I think.

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/serializer/DataFrameDescriptor.java
Comment (oleschoenburg) [reply logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/serializer/DataFrameDescriptor.java:16]: Damn, that's one of the word I never get right on the first try :D

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LoggedEventImpl.java
Comment (oleschoenburg) [reply logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LoggedEventImpl.java:139]: Re-added, although not all of the fields exist anymore so the toString is showing a bit less.

File: logstreams/src/test/java/io/camunda/zeebe/logstreams/log/SequencerTest.java
Comment (oleschoenburg) [reply logstreams/src/test/java/io/camunda/zeebe/logstreams/log/SequencerTest.java]: I added three tests here: [3dcab4a](https://github.com/camunda/zeebe/pull/11165/commits/3dcab4afeefc7d6b64dac4bde40212d42ac45d4d)

Let me know what you think

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogStorageAppender.java
Comment (oleschoenburg) [reply logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogStorageAppender.java:119]: Yeah, but I always forget about this behavior and I'd imagine I might not be the only one so I'd prefer to keep it like this, just to be explicit.

**Comment by oleschoenburg** [comment]
> Please add a unit test for the sequenced batch's serialization

I added a very basic test that serializes a batch and verifies that this is readable as multiple `LoggedEventImpl`s. Of course the test could be much more detailed (like you said: testing byte order, padding etc.) but then I'd have to re-implement large parts of the read logic just for testing and I think that our test coverage from all the other tests is good enough.

**Review by npepinpe — APPROVED** [review]
:+1: 

I think it's mostly done, not sure if we need another review. Please address the comment about testing the batched try write before merging though.

File: logstreams/src/test/java/io/camunda/zeebe/logstreams/impl/log/SequencerTest.java
Comment (oleschoenburg) [reply logstreams/src/test/java/io/camunda/zeebe/logstreams/impl/log/SequencerTest.java:27]: I didn't implement this as parameterized tests because that made them a bit difficult to follow IMO. Instead I added a companion test that writes multiple entries at once everywhere where it made sense. Hope that's okay too.

**Comment by oleschoenburg** [comment]
`StreamProcessorInconsistentPositionTest.shouldNotStartOnInconsistentLog` fails for the first time, right in time for the bors merge :sweat_smile: 
Looking into it now

**Comment by oleschoenburg** [comment]
I see. `LogContext.setupBatchWriter` needs to be adjusted.

**Review by npepinpe — APPROVED** [review]
:+1: 

I didn't see anything that really warranted another review - did I miss something? It seemed to be mostly ensuring we don't reuse records.

**Pull request merged**
